File: backend/app/routes/categorias.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List, Optional

# Imports locales
from database import get_db
from models import Categoria
from schemas import CategoriaCreate, CategoriaUpdate, CategoriaResponse

# Configuración del router
router = APIRouter(
    prefix="/categorias",
    tags=["Categorias"],
    responses={404: {"description": "No encontrado"}}
)

# ========================================
# ENDPOINTS CRUD
# ========================================

@router.get("/", response_model=List[CategoriaResponse])
def listar_categorias(
    skip: int = Query(0, ge=0, description="Registros a saltar"),
    limit: int = Query(100, ge=1, le=1000, description="Máximo registros a retornar"),
    activo: Optional[bool] = Query(None, description="Filtrar por estado activo"),
    db: Session = Depends(get_db)
):
    """
    Obtener lista de categorías con paginación
    - **skip**: Número de registros a saltar (para paginación)
    - **limit**: Máximo número de registros a retornar
    - **activo**: Filtrar solo categorías activas (true) o inactivas (false)
    """
    query = db.query(Categoria)
    if activo is not None:
        query = query.filter(Categoria.activo == activo)
    return query.offset(skip).limit(limit).all()

@router.get("/{categoria_id}", response_model=CategoriaResponse)
def obtener_categoria(
    categoria_id: int,
    db: Session = Depends(get_db)
):
    """Obtener una categoría por su ID"""
    categoria = db.query(Categoria).filter(Categoria.id_categoria == categoria_id).first()
    if categoria is None:
        raise HTTPException(status_code=404, detail="Categoría no encontrada")
    return categoria

@router.get("/codigo/{codigo}", response_model=CategoriaResponse)
def obtener_categoria_por_codigo(
    codigo: str,
    db: Session = Depends(get_db)
):
    """Obtener una categoría por su código"""
    categoria = db.query(Categoria).filter(Categoria.codigo == codigo).first()
    if categoria is None:
        raise HTTPException(status_code=404, detail=f"Categoría con código '{codigo}' no encontrada")
    return categoria

@router.post("/", response_model=CategoriaResponse, status_code=201)
def crear_categoria(
    categoria: CategoriaCreate,
    db: Session = Depends(get_db)
):
    """
    Crear una nueva categoría
    - **nombre**: Nombre de la categoría
    - **descripcion**: Descripción opcional
    - **activo**: Estado activo/inactivo
    """
    # Verificar que el código no exista
    existing_categoria = db.query(Categoria).filter(
        Categoria.codigo == categoria.codigo
    ).first()
    if existing_categoria:
        raise HTTPException(status_code=400, detail=f"Ya existe una categoría con código '{categoria.codigo}'")
    
    # Crear nueva categoría
    db_categoria = Categoria(**categoria.dict())
    db.add(db_categoria)
    db.commit()
    db.refresh(db_categoria)
    return db_categoria

@router.put("/{categoria_id}", response_model=CategoriaResponse)
def actualizar_categoria(
    categoria_id: int,
    categoria_update: CategoriaUpdate,
    db: Session = Depends(get_db)
):
    """Actualizar una categoría existente"""
    # Verificar que existe
    db_categoria = db.query(Categoria).filter(
        Categoria.id_categoria == categoria_id
    ).first()
    if not db_categoria:
        raise HTTPException(status_code=404, detail="Categoría no encontrada")
    
    # Si se está actualizando el código, verificar que no exista
    if categoria_update.codigo:
        codigo_exists = db.query(Categoria).filter(
            Categoria.codigo == categoria_update.codigo,
            Categoria.id_categoria != categoria_id
        ).first()
        if codigo_exists:
            raise HTTPException(status_code=400, detail=f"Ya existe una categoría con código '{categoria_update.codigo}'")
    
    # Actualizar campos
    update_data = categoria_update.dict(exclude_unset=True)
    for field, value in update_data.items():
        setattr(db_categoria, field, value)
    db.commit()
    db.refresh(db_categoria)
    return db_categoria

@router.delete("/{categoria_id}")
def eliminar_categoria(
    categoria_id: int,
    permanente: bool = Query(False, description="Eliminación permanente (true) o soft delete (false)"),
    db: Session = Depends(get_db)
):
    """
    Eliminar una categoría
    - **permanente**: Si es true, elimina permanentemente. Si es false, solo marca como inactivo.
    """
    db_categoria = db.query(Categoria).filter(
        Categoria.id_categoria == categoria_id
    ).first()
    if not db_categoria:
        raise HTTPException(status_code=404, detail="Categoría no encontrada")
    
    if permanente:
        db.delete(db_categoria)
        message = "Categoría eliminada permanentemente"
    else:
        db_categoria.activo = False
        message = "Categoría desactivada"
    db.commit()
    return {"message": message}

@router.patch("/{categoria_id}/activar")
def activar_categoria(
    categoria_id: int,
    db: Session = Depends(get_db)
):
    """Activar una categoría (cambiar activo = true)"""
    db_categoria = db.query(Categoria).filter(
        Categoria.id_categoria == categoria_id
    ).first()
    if not db_categoria:
        raise HTTPException(status_code=404, detail="Categoría no encontrada")
    
    db_categoria.activo = True
    db.commit()
    db.refresh(db_categoria)
    return {"message": "Categoría activada", "categoria": db_categoria}

# ========================================
# ENDPOINTS ADICIONALES
# ========================================

@router.get("/estadisticas/resumen")
def obtener_estadisticas_categorias(db: Session = Depends(get_db)):
    """Obtener estadísticas generales de categorías"""
    total = db.query(Categoria).count()
    activas = db.query(Categoria).filter(Categoria.activo == True).count()
    inactivas = total - activas
    return {
        "total": total,
        "activas": activas,
        "inactivas": inactivas,
        "porcentaje_activas": round((activas / total * 100) if total > 0 else 0, 2)
    }

@router.post("/bulk", response_model=List[CategoriaResponse])
def crear_categorias_masivo(
    categorias: List[CategoriaCreate],
    db: Session = Depends(get_db)
):
    """Crear múltiples categorías de una vez"""
    if len(categorias) > 50:
        raise HTTPException(status_code=400, detail="Máximo 50 categorías por operación masiva")
    
    # Verificar códigos duplicados en la lista
    codigos = [c.codigo for c in categorias]
    if len(codigos) != len(set(codigos)):
        raise HTTPException(status_code=400, detail="Hay códigos duplicados en la lista")
    
    # Verificar códigos existentes en BD
    existing_codes = db.query(Categoria.codigo).filter(
        Categoria.codigo.in_(codigos)
    ).all()
    if existing_codes:
        codes_str = ", ".join([code[0] for code in existing_codes])
        raise HTTPException(status_code=400, detail=f"Los siguientes códigos ya existen: {codes_str}")
    
    # Crear todas las categorías
    db_categorias = []
    for categoria_data in categorias:
        db_categoria = Categoria(**categoria_data.dict())
        db.add(db_categoria)
        db_categorias.append(db_categoria)
    db.commit()
    
    # Refresh todas las categorías
    for db_categoria in db_categorias:
        db.refresh(db_categoria)
    
    return db_categorias

File: frontend/src/stores/categorias.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { useApiStore } from './api'

export interface Categoria {
  id_categoria: number
  codigo_categoria: string
  nombre_categoria: string
  descripcion?: string
  activo: boolean
  subcategorias?: Subcategoria[]
}

export interface CategoriaCreate {
  codigo_categoria: string
  nombre_categoria: string
  descripcion?: string
  activo: boolean
}

export interface CategoriaUpdate {
  codigo_categoria?: string
  nombre_categoria?: string
  descripcion?: string
  activo?: boolean
}

export interface Subcategoria {
  id_subcategoria: number
  id_categoria: number
  codigo_subcategoria: string
  nombre_subcategoria: string
  descripcion?: string
  activo: boolean
  categoria?: Categoria
}

export interface SubcategoriaCreate {
  id_categoria: number
  codigo_subcategoria: string
  nombre_subcategoria: string
  descripcion?: string
  activo: boolean
}

export interface SubcategoriaUpdate {
  id_categoria?: number
  codigo_subcategoria?: string
  nombre_subcategoria?: string
  descripcion?: string
  activo?: boolean
}

export const useCategoriaStore = defineStore('categorias', () => {
  const apiStore = useApiStore()

  const categorias = ref<Categoria[]>([])
  const subcategorias = ref<Subcategoria[]>([])
  const isLoading = ref(false)

  // CRUD Categorías
  const obtenerCategorias = async (params?: {
    skip?: number
    limit?: number
    activo?: boolean
  }): Promise<Categoria[]> => {
    try {
      isLoading.value = true
      const queryParams = new URLSearchParams()

      if (params?.skip !== undefined) queryParams.append('skip', params.skip.toString())
      if (params?.limit !== undefined) queryParams.append('limit', params.limit.toString())
      if (params?.activo !== undefined) queryParams.append('activo', params.activo === true ? 'true' : 'false')

      const url = `/categorias?${queryParams.toString()}`
      console.log('URL construida:', url, 'params recibidos:', params)
      const response = await apiStore.get(url)
      if (!response.success) {
        throw new Error(response.error || 'Error al obtener categorías')
      }
      categorias.value = response.data
      return response.data
    } catch (error) {
      console.error('Error al obtener categorías:', error)
      throw error
    } finally {
      isLoading.value = false
    }
  }

  const obtenerCategoria = async (id: number): Promise<Categoria> => {
    try {
      const response = await apiStore.get(`/categorias/${id}`)
      if (!response.success) {
        throw new Error(response.error || 'Error al obtener categoría')
      }
      return response.data
    } catch (error) {
      console.error('Error al obtener categoría:', error)
      throw error
    }
  }

  const obtenerCategoriaPorCodigo = async (codigo: string): Promise<Categoria> => {
    try {
      const response = await apiStore.get(`/categorias/codigo/${codigo}`)
      if (!response.success) {
        throw new Error(response.error || 'Error al obtener categoría por código')
      }
      return response.data
    } catch (error) {
      console.error('Error al obtener categoría por código:', error)
      throw error
    }
  }

  const crearCategoria = async (categoria: CategoriaCreate): Promise<Categoria> => {
    try {
      const response = await apiStore.post('/categorias', categoria)
      if (!response.success) {
        throw new Error(response.error || 'Error al crear categoría')
      }
      return response.data
    } catch (error) {
      console.error('Error al crear categoría:', error)
      throw error
    }
  }

  const actualizarCategoria = async (id: number, categoria: CategoriaUpdate): Promise<Categoria> => {
    try {
      const response = await apiStore.put(`/categorias/${id}`, categoria)
      if (!response.success) {
        throw new Error(response.error || 'Error al actualizar categoría')
      }
      return response.data
    } catch (error) {
      console.error('Error al actualizar categoría:', error)
      throw error
    }
  }

  const eliminarCategoria = async (id: number, permanente: boolean = false): Promise<void> => {
    try {
      const params = new URLSearchParams()
      params.append('permanente', permanente.toString())

      const response = await apiStore.delete(`/categorias/${id}?${params.toString()}`)
      if (!response.success) {
        throw new Error(response.error || 'Error al eliminar categoría')
      }
    } catch (error) {
      console.error('Error al eliminar categoría:', error)
      throw error
    }
  }

  const activarCategoria = async (id: number): Promise<Categoria> => {
    try {
      const response = await apiStore.patch(`/categorias/${id}/activar`)
      if (!response.success) {
        throw new Error(response.error || 'Error al activar categoría')
      }
      return response.data.categoria
    } catch (error) {
      console.error('Error al activar categoría:', error)
      throw error
    }
  }

  const obtenerEstadisticasCategorias = async () => {
    try {
      const response = await apiStore.get('/categorias/estadisticas/resumen')
      if (!response.success) {
        throw new Error(response.error || 'Error al obtener estadísticas de categorías')
      }
      return response.data
    } catch (error) {
      console.error('Error al obtener estadísticas de categorías:', error)
      throw error
    }
  }

  const crearCategoriasMasivo = async (categorias: CategoriaCreate[]): Promise<Categoria[]> => {
    try {
      const response = await apiStore.post('/categorias/bulk', categorias)
      if (!response.success) {
        throw new Error(response.error || 'Error al crear categorías masivo')
      }
      return response.data
    } catch (error) {
      console.error('Error al crear categorías masivo:', error)
      throw error
    }
  }

  // CRUD Subcategorías
  const obtenerSubcategorias = async (params?: {
    skip?: number
    limit?: number
    activo?: boolean
  }): Promise<Subcategoria[]> => {
    try {
      isLoading.value = true
      const queryParams = new URLSearchParams()

      if (params?.skip !== undefined) queryParams.append('skip', params.skip.toString())
      if (params?.limit !== undefined) queryParams.append('limit', params.limit.toString())
      if (params?.activo !== undefined) queryParams.append('activo', params.activo.toString())

      const response = await apiStore.get(`/subcategorias?${queryParams.toString()}`)
      if (!response.success) {
        throw new Error(response.error || 'Error al obtener subcategorías')
      }
      subcategorias.value = response.data
      return response.data
    } catch (error) {
      console.error('Error al obtener subcategorías:', error)
      throw error
    } finally {
      isLoading.value = false
    }
  }

  const obtenerSubcategoria = async (id: number): Promise<Subcategoria> => {
    try {
      const response = await apiStore.get(`/subcategorias/${id}`)
      if (!response.success) {
        throw new Error(response.error || 'Error al obtener subcategoría')
      }
      return response.data
    } catch (error) {
      console.error('Error al obtener subcategoría:', error)
      throw error
    }
  }

  const obtenerSubcategoriaPorCodigo = async (codigo: string): Promise<Subcategoria> => {
    try {
      const response = await apiStore.get(`/subcategorias/codigo/${codigo}`)
      if (!response.success) {
        throw new Error(response.error || 'Error al obtener subcategoría por código')
      }
      return response.data
    } catch (error) {
      console.error('Error al obtener subcategoría por código:', error)
      throw error
    }
  }

  const obtenerSubcategoriasPorCategoria = async (
    idCategoria: number,
    params?: {
      skip?: number
      limit?: number
      activo?: boolean
    }
  ): Promise<Subcategoria[]> => {
    try {
      const queryParams = new URLSearchParams()

      if (params?.skip !== undefined) queryParams.append('skip', params.skip.toString())
      if (params?.limit !== undefined) queryParams.append('limit', params.limit.toString())
      if (params?.activo !== undefined) queryParams.append('activo', params.activo.toString())

      const response = await apiStore.get(`/subcategorias/por-categoria/${idCategoria}?${queryParams.toString()}`)
      if (!response.success) {
        throw new Error(response.error || 'Error al obtener subcategorías por categoría')
      }
      return response.data
    } catch (error) {
      console.error('Error al obtener subcategorías por categoría:', error)
      throw error
    }
  }

  const crearSubcategoria = async (subcategoria: SubcategoriaCreate): Promise<Subcategoria> => {
    try {
      const response = await apiStore.post('/subcategorias', subcategoria)
      if (!response.success) {
        throw new Error(response.error || 'Error al crear subcategoría')
      }
      return response.data
    } catch (error) {
      console.error('Error al crear subcategoría:', error)
      throw error
    }
  }

  const actualizarSubcategoria = async (id: number, subcategoria: SubcategoriaUpdate): Promise<Subcategoria> => {
    try {
      const response = await apiStore.put(`/subcategorias/${id}`, subcategoria)
      if (!response.success) {
        throw new Error(response.error || 'Error al actualizar subcategoría')
      }
      return response.data
    } catch (error) {
      console.error('Error al actualizar subcategoría:', error)
      throw error
    }
  }

  const eliminarSubcategoria = async (id: number, permanente: boolean = false): Promise<void> => {
    try {
      const params = new URLSearchParams()
      params.append('permanente', permanente.toString())

      const response = await apiStore.delete(`/subcategorias/${id}?${params.toString()}`)
      if (!response.success) {
        throw new Error(response.error || 'Error al eliminar subcategoría')
      }
    } catch (error) {
      console.error('Error al eliminar subcategoría:', error)
      throw error
    }
  }

  const activarSubcategoria = async (id: number): Promise<Subcategoria> => {
    try {
      const response = await apiStore.patch(`/subcategorias/${id}/activar`)
      if (!response.success) {
        throw new Error(response.error || 'Error al activar subcategoría')
      }
      return response.data.subcategoria
    } catch (error) {
      console.error('Error al activar subcategoría:', error)
      throw error
    }
  }

  const obtenerEstadisticasSubcategorias = async () => {
    try {
      const response = await apiStore.get('/subcategorias/estadisticas/resumen')
      if (!response.success) {
        throw new Error(response.error || 'Error al obtener estadísticas de subcategorías')
      }
      return response.data
    } catch (error) {
      console.error('Error al obtener estadísticas de subcategorías:', error)
      throw error
    }
  }

  const crearSubcategoriasMasivo = async (subcategorias: SubcategoriaCreate[]): Promise<Subcategoria[]> => {
    try {
      const response = await apiStore.post('/subcategorias/bulk', subcategorias)
      if (!response.success) {
        throw new Error(response.error || 'Error al crear subcategorías masivo')
      }
      return response.data
    } catch (error) {
      console.error('Error al crear subcategorías masivo:', error)
      throw error
    }
  }

  return {
    // State
    categorias,
    subcategorias,
    isLoading,

    // Categorías
    obtenerCategorias,
    obtenerCategoria,
    obtenerCategoriaPorCodigo,
    crearCategoria,
    actualizarCategoria,
    eliminarCategoria,
    activarCategoria,
    obtenerEstadisticasCategorias,
    crearCategoriasMasivo,

    // Subcategorías
    obtenerSubcategorias,
    obtenerSubcategoria,
    obtenerSubcategoriaPorCodigo,
    obtenerSubcategoriasPorCategoria,
    crearSubcategoria,
    actualizarSubcategoria,
    eliminarSubcategoria,
    activarSubcategoria,
    obtenerEstadisticasSubcategorias,
    crearSubcategoriasMasivo
  }
})